tracing/trace_state.hh

```diff
@@ -477,14 +477,16 @@ private:
      * @param fmt format string
      * @param a positional parameters
      */
-    template <typename... A>
-    void trace(const char* fmt, A&&... a) noexcept;
+    template <typename... T>
+    void trace(fmt::format_string<T...> fmt, T&&... args) noexcept;
 
     template <typename... A>
     friend void begin(const trace_state_ptr& p, A&&... a);
 
-    template <typename... A>
-    friend void trace(const trace_state_ptr& p, A&&... a) noexcept;
+    template <typename... T>
+    friend void trace(const trace_state_ptr& p, fmt::format_string<T...>, T&&... args) noexcept;
+
+    friend void trace(const trace_state_ptr& p, const sstring& msg) noexcept;
```

Comment:
This one would be matched with `fmt::format_string`. Bu we cannot detect something like:

```c++
string hey = "hello {}";
trace(trace_ptr, hey);
```

Unless this usage is forbidden — the same dilemma we are facing in seastar’s pr.

Reply:
I think the second overload is more favorable than the first, because the first is a template. So C++ would prefer converting "foo {} bar" to std::string rather than fmt::format_string.

In Seastar we have to deal with compatibility, but here we don't, we can just require `trace(trace_ptr, "{}", hey);` .

Reply:
ahh, you are right. see https://godbolt.org/z/czTjj49xc . i am dropping this overload in a new patch.

Reply:
#14188